uaclient/apt.py

```diff
@@ -22,36 +23,37 @@
 CA_CERTIFICATES_FILE = '/usr/sbin/update-ca-certificates'
 
 
-class InvalidAPTCredentialsError(RuntimeError):
-    """Raised when invalid token is provided for APT access"""
-    pass
-
-
-def valid_apt_credentials(repo_url, username, password):
+def assert_valid_apt_credentials(repo_url, username, password):
     """Validate apt credentials for a PPA.
 
     @param repo_url: private-ppa url path
     @param username: PPA login username.
     @param password: PPA login password or resource token.
 
-    @return: True if valid or unable to validate
+    @raises: UserFacingError for invalid credentials, timeout or unexpected
+        errors.
     """
     protocol, repo_path = repo_url.split('://')
     if not os.path.exists('/usr/lib/apt/apt-helper'):
-        return True   # Do not validate
+        return
     try:
         util.subp(['/usr/lib/apt/apt-helper', 'download-file',
                    '%s://%s:%s@%s/ubuntu/pool/' % (
                        protocol, username, password, repo_path),
-                   '/tmp/uaclient-apt-test'],
-                  capture=False)  # Hide credentials from logs
-        return True
-    except util.ProcessExecutionError:
-        pass
+                   '/tmp/uaclient-apt-test'])
+    except util.ProcessExecutionError as e:
+        if e.exit_code == 100:
+            stderr = str(e.stderr).lower()
+            if re.search(r'401\s+unauthorized|httperror401', stderr):
+                raise exceptions.UserFacingError(
+                    'Invalid APT credentials provided for %s' % repo_url)
+            elif re.search(r'connection timed out', stderr):
+                raise exceptions.UserFacingError(
+                    'Timeout trying to access APT repository at %s' % repo_url)
+        raise exceptions.UserFacingError(str(e))
```

Comment:
What is `str(e)` going to look like here?  If it's more than a single-line summary then I don't think we should pipe that straight out to the user.

Reply:
(FAOD, and to save another round-trip, if it is more verbose than we want, I think we should just emit "unknown apt problem sorry" and people can reference the logs or run with --debug to get more.)

Reply:
Your assumption is correct here. other errors are a bit too long for taste.

ERROR: Failed running command '/usr/lib/apt/apt-helper2 download-file https://bearer:mAgJhAwoQ4Hs8wpIhS-nr72eEdrg9gBIBMBpJCkRtdC0AAMpzfzfXBQmgTvoLGE9-Rynn_lNCNulp4v_VvVoaiD05YWEzMjlhZWE5YTc4NGUzNGUzN2NiYjQ1Y2ZmYzRmNBIBKgAABiCbND3Clmt8PKmmTgG_eYiwLbqzOKouJsIDrnr7rmfGsg@esm.staging.ubuntu.com/ubuntu/pool/ /tmp/uaclient-apt-test' [exit(1)]. Message: b'my error string\n'